sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx

```diff
@@ -0,0 +1,192 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# cython: profile=True
+
+cimport cython
+from apache_beam.utils.counters import Counter
+from apache_beam.utils.counters import CounterName
+
+from cython.operator cimport dereference as dref
+from cython.operator cimport preincrement as iref
+
+cdef class Journal(object):
+  def __init__(self):
+    self.max_snapshot = -1
+    self.journal = deque[SnapshottedExecutionPtr]()
+
+  cdef void add_journal(self, ElementExecutionPtr execution_ptr,
+                        int64_t snapshot):
+    if snapshot <= self.max_snapshot:
```

Comment:
<!--new_thread; commit:1f7d05e01de972b0e3bec788fe9c82377ea44f15; resolved:0-->
It seems that it should be called min_snapshot? : )

Reply:
Actually it's max_snapshot since journal keeps track of the increasing snapshot number and always keep the largest one.